### PR TITLE
Git 39

### PR DESCRIPTION
**Pull request by SaraPopa**

### CODE REVIEW

File: lib/functions/losses.ts [hunk 1]
```diff
@@ -16,7 +16,7 @@ function meanSquaredError(prediction: number[], target: number[]): number {
         value += (prediction[i] - target[i]) ** 2
     }
 
-    return value / target.length
+    return value / (2 * target.length)
```
Comment (cstefanache) [inline lib/functions/losses.ts:19]: why 2*?

File: lib/regressions/linearRegression.ts [hunk 2]
```diff
@@ -1,50 +1,34 @@
...
-    for (let i = 0; i < input.length; i++) {
-        value.push(weight * input[i] + bias)
-    }
-
-    return value
-}
+import { VectorUtils } from '../utils/math-utils'
+import predictSinglevariable, { RegressionOutputType } from './utilities'
 
 export default class LinearRegression {
```
Comment (cstefanache) [inline lib/regressions/linearRegression.ts:6]: should extend evolutionary algorithm (see clustering)

File: lib/regressions/multivariableLinearRegression.ts [hunk 3]
```diff
@@ -1,55 +1,33 @@
...
-            pred += feature[i] * weight[i]
-        }
-        value.push(pred + bias)
-    })
-
-    return value
-}
+import { predictMultivariable, RegressionOutputType, transposeAndNormalize } from './utilities'
 
 export default class MultivariableLinearRegression {
```
Comment (cstefanache) [inline lib/regressions/multivariableLinearRegression.ts:5]: also can extend evolutionary algorithm

File: lib/regressions/multivariableLinearRegression.ts [hunk 4]
```diff
@@ -1,55 +1,33 @@
...
         const costHistory = []
-        let updatedWeight = weight
-        let updatedBias = bias
-        let updatedPrediction = predictionMultivariable(input, weight, bias)
+        const transformedInput = transposeAndNormalize(input)
+        let biasAndWeights = Array(input[0].length + 1).fill(0)
         let currentEpoch = 0
+        let updatedPrediction = predictMultivariable(transformedInput, biasAndWeights)
 
         while (true) {
```
Comment (cstefanache) [inline lib/regressions/multivariableLinearRegression.ts:19]: is there a stopping condition?

File: lib/regressions/polynomialRegression.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            updatedPrediction = predictMultivariable(transformedInput, biasAndWeights)
+
+            currentEpoch += 1
+
+            if (
+                costHistory.length >= 3 &&
+                costHistory[costHistory.length - 1] === costHistory[costHistory.length - 2] &&
+                costHistory[costHistory.length - 2] === costHistory[costHistory.length - 3]
+            ) {
+                updated = false
```
Comment (cstefanache) [inline lib/regressions/polynomialRegression.ts:45]: If the algorithm extends an evolutionary algorithm a Convergence instance can be used to signal stop condition - check KMeans implementation

File: lib/functions/losses.ts (hunk 1 above)
Comment (SaraPopa) [reply lib/functions/losses.ts:19]: Dividing the expression by 2 will make the derivative calculations simpler. 
From `-2 * input * (target - prediction)` to `-1 * input * (target - prediction)`

File: lib/regressions/multivariableLinearRegression.ts (hunk 4 above)
Comment (SaraPopa) [reply lib/regressions/multivariableLinearRegression.ts:19]: Should there be a stopping condition? There I initialize the variable `updatedPrediction`.

**Pull request merged**